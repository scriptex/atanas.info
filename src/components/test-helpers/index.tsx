import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { render, RenderResult, waitFor } from '@testing-library/react';

import * as shared from '../../scripts/shared';

export const mockUseEffect = (): void => {
	let useEffect: jest.Mock<React.EffectCallback>;

	const mockedUseEffect = () => {
		useEffect.mockImplementationOnce((f: () => React.EffectCallback) => f());
	};

	beforeEach(() => {
		useEffect = jest.spyOn(React, 'useEffect') as unknown as jest.Mock<React.EffectCallback>;

		mockedUseEffect();
	});
};

export const snapshotTest = (
	Component: React.FC<Readonly<React.ComponentProps<any>>>,
	buttonSelector?: string
): void => {
	const { name } = Component;

	jest.spyOn(shared, 'formatDate');

	(shared.formatDate as jest.Mock).mockImplementation(() => 'Mock date');

	mockUseEffect();

	afterAll(() => {
		jest.clearAllMocks();
	});

	describe(`${name} component`, () => {
		it(`Should render the ${name} component`, async () => {
			let result!: RenderResult;

			await act(async () => {
				result = await waitFor(() => render(<Component />));
			});

			const { asFragment, container } = result;

			expect(asFragment()).toMatchSnapshot();

			if (buttonSelector) {
				act(() => {
					container.querySelectorAll<HTMLElement>(buttonSelector).forEach(button => {
						if (!button.hasAttribute('download')) {
							button.click();

							expect(asFragment()).toMatchSnapshot();
						}
					});
				});
			}
		});

		it(`Should unmount the ${name} component`, async () => {
			let result!: RenderResult;

			await act(async () => {
				result = await waitFor(() => render(<Component />));
			});

			const { unmount, asFragment } = result;

			unmount();

			expect(asFragment()).toMatchSnapshot();
		});
	});
};
