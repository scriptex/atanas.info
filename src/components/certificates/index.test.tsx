import { Certificates } from '.';
import { snapshotTest } from '../test-helpers';

snapshotTest(Certificates);
