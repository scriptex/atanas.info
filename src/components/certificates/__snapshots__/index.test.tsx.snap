// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Certificates component Should render the Certificates component 1`] = `
<DocumentFragment>
  <main
    class="o-main "
  >
    <section
      class="c-section c-section--certificates"
      id="certificates"
    >
      <div
        class="o-shell"
      >
        <header
          class="c-section__header"
        >
          <div
            class="c-section__animation"
            style="width: 150px; height: 150px; margin-left: auto; margin-right: auto;"
          />
          <h2>
            Certificates
          </h2>
        </header>
        <div
          class="c-section__body o-grid"
        >
          <div
            class="o-grid__item xs-12 sm-6"
          >
            <a
              class="c-certificate"
              href="cert/html.pdf"
              rel="noopener noreferrer"
              target="_blank"
            >
              <div
                class="c-loading"
              >
                <div
                  class="c-loader"
                />
              </div>
              <img
                alt="HTML"
                loading="lazy"
                src="cert/html.jpg"
                width="600"
              />
            </a>
          </div>
          <div
            class="o-grid__item xs-12 sm-6"
          >
            <a
              class="c-certificate"
              href="cert/javascript.pdf"
              rel="noopener noreferrer"
              target="_blank"
            >
              <div
                class="c-loading"
              >
                <div
                  class="c-loader"
                />
              </div>
              <img
                alt="JavaScript"
                loading="lazy"
                src="cert/javascript.jpg"
                width="600"
              />
            </a>
          </div>
          <div
            class="o-grid__item xs-12 sm-6"
          >
            <a
              class="c-certificate"
              href="cert/python.pdf"
              rel="noopener noreferrer"
              target="_blank"
            >
              <div
                class="c-loading"
              >
                <div
                  class="c-loader"
                />
              </div>
              <img
                alt="Python"
                loading="lazy"
                src="cert/python.jpg"
                width="600"
              />
            </a>
          </div>
          <div
            class="o-grid__item xs-12 sm-6"
          >
            <a
              class="c-certificate"
              href="cert/react.pdf"
              rel="noopener noreferrer"
              target="_blank"
            >
              <div
                class="c-loading"
              >
                <div
                  class="c-loader"
                />
              </div>
              <img
                alt="React Development"
                loading="lazy"
                src="cert/react.jpg"
                width="600"
              />
            </a>
          </div>
          <div
            class="o-grid__item xs-12 sm-6"
          >
            <a
              class="c-certificate"
              href="cert/css.pdf"
              rel="noopener noreferrer"
              target="_blank"
            >
              <div
                class="c-loading"
              >
                <div
                  class="c-loader"
                />
              </div>
              <img
                alt="CSS"
                loading="lazy"
                src="cert/css.jpg"
                width="600"
              />
            </a>
          </div>
          <div
            class="o-grid__item xs-12 sm-6"
          >
            <a
              class="c-certificate"
              href="cert/deno.pdf"
              rel="noopener noreferrer"
              target="_blank"
            >
              <div
                class="c-loading"
              >
                <div
                  class="c-loader"
                />
              </div>
              <img
                alt="Deno"
                loading="lazy"
                src="cert/deno.jpg"
                width="600"
              />
            </a>
          </div>
        </div>
        <div
          class="c-section__actions"
        >
          <button
            class="c-btn "
            type="button"
          >
            Get in touch
          </button>
          <ul
            class="c-section__actions-list"
          >
            <li>
              <button
                class="c-btn c-section__actions-close"
                type="button"
              >
                Close
              </button>
            </li>
            <li>
              <button
                class="c-btn "
                type="button"
              >
                <svg
                  class="c-section__actions-icon"
                >
                  <use
                    xlink:href="#svg-email"
                  />
                </svg>
                <span>
                  Contact me
                </span>
                <strong>
                  directly
                </strong>
              </button>
            </li>
            <li>
              <a
                class="c-btn"
                href="https://app.usebraintrust.com/talent/782/"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  class="c-section__actions-icon"
                >
                  <use
                    xlink:href="#svg-braintrust"
                  />
                </svg>
                <span>
                  Hire me on
                </span>
                <strong>
                  Braintrust
                </strong>
              </a>
            </li>
            <li>
              <a
                class="c-btn"
                href="https://www.toptal.com/resume/atanas-atanasov"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  class="c-section__actions-icon"
                >
                  <use
                    xlink:href="#svg-toptal"
                  />
                </svg>
                <span>
                  Hire me on
                </span>
                <strong>
                  Toptal
                </strong>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </section>
  </main>
</DocumentFragment>
`;
